blueprint/core/blueprint_EcmascriptEngine.cpp

```diff
@@ -107,6 +135,31 @@ namespace blueprint
 
         // Add console.log support
         duk_console_init(ctx, DUK_CONSOLE_FLUSH);
+
+        // Create our error handler wrapping a user supplied error callback.
+        errorHandler = [=]
+        {
+            if (onUncaughtError == nullptr)
+                duk_throw_raw(ctx);
+
+            const juce::String trace = duk_safe_to_stacktrace(ctx, -1);
+            const juce::String msg = duk_safe_to_string(ctx, -1);
+
+            // Call the user provided error handler
+            std::invoke(onUncaughtError, msg, trace);
+
+            // Clear out the stack so we can re-register native functions
+            // after we clear out the lambda release pool etc.
+            while (duk_get_top(ctx))
+            {
+                duk_remove(ctx, duk_get_top_index(ctx));
+            }
+
+            // Clear the LambdaHelper release pool as duktape does not call object
+            // finalizers in the event of an evaluation error or duk_pcall failure.
+            lambdaReleasePool.clear();
```

Comment:
Hm, I must have missed this in the earlier discussion, but if an error is successfully caught and handled, why do we have to clear the lambda pool and the duk stack?

Reply:
So, as far as I can see duktape didn't call any finalizers when we reload the bundle and an evaluation errors occurs, meaning the size of the release pool may steadily increases. This may have been antagonised by what appears to be a separate problem with the `CanvasView`, with `animate` enabled the UI eventually locks up when just sat idle as there seems to be some delay garbage collection all the intermediate function objects associated with the `CanvasContext`.  That seems to result in the release poll getting huge and then when it is finally clears it blocks the UI for an age. With animate enabled under a hot-reload workflow this got even worse when there were evaluation errors as things weren't getting cleaned up. 

Also, in the event of a bundle error we may reload our bundle in a second run (after fixing an error), causing `beforeBundleEval` to be called. This results in duplicating all native props and functions on the duktape stack because we/duktape does not clear the stack of these items in the event of an error. 

This may require more scrutiny but this approach provides consistent results. i.e. when there is an error we "reset" the `EcmascriptEngine`. This requires callers fixup the bundle and reload (or if the error is in C++ they'll have to rebuild). 

Arguably we should only clear out the EcmascriptEngine when the error is an eval error on a bundle. It might make more sense to give `EcmascriptEngine` a `reset` function or something which `ReactApplicationRoot` can call under specific error cases. 

Not totally sure what the best approach is here.

Reply:
Ahh interesting, I see what you mean. I agree, there's some nuance here. For example:

```js
function a() {
  try {
    return 2 + doesntExist();
  } catch (e) {
    console.log('caught it');
  }
}

a();
```

if I evaluate the above in a javascript environment, `a` is still defined, and can be invoked again. We could imagine removing that try/catch block from within `a` here and instead catching that at the cpp level with our new approach to error handling:

```cpp
try {
    engine.invoke("a");
} catch (EcmascriptEngine::Error const& e) {
    DBG("Caught it");
}
```

Here I think we might stilll expect `a` to be defined in the engine and able to invoke it again, as in the case above. Now what if this function `a` is actually a native method with an attached lambda in the pool? This case would be a problem.

That said, I do think there's a lot of nuance here and I also think this PR has grown already to take on many problems :) Let's go ahead with this part as is so that we can lock in debugging, and we can continue the conversation on error handling + implementation in a separate issue. Sound good?

Reply:
Sounds good to me @nick-thompson. Maybe we should raise a "Refine error handling" issue or similar to cover this and I'll drop over an initial PR in the next couple of days with some ideas.

Reply:
FYI, I've tested current state of this branch on all 3 platforms and everything is good. We'll just need to add Debug setup documentation to the WIKI at some point. Need to add a section detailing that Windows currently requires a hack to the vscode extension. We could just link to a branch on my own fork or something for now.

Reply:
Yea, perfect!

Reply:
Cool. I'll get that WIKI page finished off an drop you a message or something when it's done and you can grab from my fork. Annoying you can't PR a wiki page!

Reply:
Sounds great, thanks. And yea, annoying indeed... there must be a way to make that easier